src/app/clusters/on-off-server/on-off-server.cpp

```diff
@@ -48,7 +48,9 @@ using chip::Protocols::InteractionModel::Status;
 static OnOffEffect * firstEffect = nullptr;
 OnOffServer OnOffServer::instance;
 
-static EmberEventControl gEventControls[EMBER_AF_ON_OFF_CLUSTER_SERVER_ENDPOINT_COUNT];
+static constexpr size_t kOnOffMaxEnpointCount =
+    EMBER_AF_ON_OFF_CLUSTER_SERVER_ENDPOINT_COUNT + CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT;
+static EmberEventControl gEventControls[kOnOffMaxEnpointCount];
```

Comment:
This is not enough on its own.

The array is indexed into using `emberAfFindClusterServerEndpointIndex`.  But that function is only safe to use for static endpoints.  For dynamic endpoints it can at runtime change what value it returns for the same exact endpoint.

See also the description of this same problem in https://github.com/project-chip/connectedhomeip/issues/25984#issue-1655772149

We should probably have a helper function for this "find the index into the array for both static and dynamic endpoints" operation....

And we ideally should have tests for this.

Reply:
Ok I didn't realize that emberAfFindClusterServerEndpointIndex was not properly working for dynamic endpoints. 

Why don't we set dynamic clusters to always be set at the end of the static indexes?
let's say I have `EMBER_AF_ON_OFF_CLUSTER_SERVER_ENDPOINT_COUNT` static on-off cluster defined. Then, a dynamic cluster added will always be set to an index > than  `(EMBER_AF_ON_OFF_CLUSTER_SERVER_ENDPOINT_COUNT-1)`

Reply:
if I look at the implementation of `emberAfGetDynamicIndexFromEndpoint` and `emberAfSetDynamicEndpoint`it seems to be implemented as I mention. What am I missing? why does the index of a dynamic endpoint change at runtime?

Reply:
@jmartinez-silabs `emberAfFindClusterServerEndpointIndex` is the index of the endpoint in the list of endpoints that implement the given cluster, and with dynamic endpoints that list can have things added/removed in the middle as dynamic endpoints are added/removed.

Let's say you have a device with one static endpoint, with endpoint id 1, that implements On/Off and allows two dynamic endpoints, and the following sequence of actions happens:

1. A dynamic endpoint with endpoint id 2 is defined at dynamic index 0 that implements On/Off.
2. A dynamic endpoint with endpoint id 3 is defined at dynamic index 1 that implements On/Off.

At this point:

* `emberAfFindClusterServerEndpointIndex(1, OnOff::Id) == 0`
* `emberAfFindClusterServerEndpointIndex(2, OnOff::Id) == 1`
* `emberAfFindClusterServerEndpointIndex(3, OnOff::Id) == 2`

Now endpoint 2 is removed via `emberAfClearDynamicEndpoint(0)` (e.g. because we are a bridge and the device we are bridging is permanently disconnected from the bridge.  Now what we have is:

* `emberAfFindClusterServerEndpointIndex(1, OnOff::Id) == 0`
* `emberAfFindClusterServerEndpointIndex(3, OnOff::Id) == 1`

Now we start bridging a new thing, endpoint 4 is added at dynamic index 0, and we end up with:

* `emberAfFindClusterServerEndpointIndex(1, OnOff::Id) == 0`
* `emberAfFindClusterServerEndpointIndex(4, OnOff::Id) == 1`
* `emberAfFindClusterServerEndpointIndex(3, OnOff::Id) == 2`

Reply:
Do you agree that when we remove the dynamic endpoint 2, we should not re-index the existing dynamic endpoints. In my opinion, we should leave index 1 empty and the 2nd dynamic endpoint(3) keeps his initial his index (2 in the case above). 

This way you would always be sure that for a given existing dynamic endpoint you always get the same index. 

I am not sure how we can identify a dynamic ep with his index with what is currently implemented

Reply:
> Do you agree that when we remove the dynamic endpoint 2, we should not re-index the existing dynamic endpoints.

"re-index" in what sense?  Their _endpoint index_ does not change. But the value returned by `emberAfFindClusterServerEndpointIndex` changes, as things stand.

We could change what `emberAfFindClusterServerEndpointIndex` does, of course.  What is the exact proposal for what it should do?

> I am not sure how we can identify a dynamic ep with his index with what is currently implemented

You can identify dynamic endpoints by:

1. `emberAfIndexFromEndpoint(endpoint_id)` is not `kEmberInvalidEndpointIndex`
2. `emberAfIndexFromEndpoint(endpoint_id) >= emberAfFixedEndpointCount()`

Reply:
I added a new function that returns the index of the ClusterServer endpoint as it is indexed in `emAfEndpoints`.

This index is always valid for static and dynamic endpoints and will correspond to the event control array on the on-off server implementation.

This new function can be used in multiple cluster server implementations and fix the issue with dynamic endpoint indexing when using `emberAfFindClusterServerEndpointIndex`.

Once this is merged, I'll replace the function calls in the other cluster that use a similar implementation than OnOff cluster